File: src/main/java/com/fireblazerrr/survivorbot/channel/Channel.java
package com.fireblazerrr.survivorbot.channel;

import com.fireblazerrr.survivorbot.chatter.Chatter;
import com.fireblazerrr.survivorbot.spigot.events.ChannelChatEvent;
import com.fireblazerrr.survivorbot.utils.message.MessageFormatSupplier;
import net.md_5.bungee.api.chat.TextComponent;
import org.bukkit.ChatColor;
import org.bukkit.World;
import org.bukkit.entity.Player;

import java.util.Set;

public interface Channel {

    boolean addMember(Chatter chatter, boolean announce, boolean flag);

    void addWorld(String world);

    void announce(String message);

    void announce(TextComponent textComponent);

    TextComponent applyFormat(String format, String msg, Player player);

    String applyFormat(String format, String originalFormat);

    boolean banMember(Chatter chatter, boolean bool, boolean flag);

    void emote(Chatter chatter, String message);

    Set<String> getBans();

    void setBans(Set<String> bans);

    ChatColor getColor();

    void setColor(ChatColor color);

    int getDistance();

    void setDistance(int distance);

    String getFormat();

    void setFormat(String format);

    Set<Chatter> getMembers();

    Set<String> getModerators();

    void setModerators(Set<String> chatters);

    Set<String> getMutes();

    void setMutes(Set<String> chatters);

    String getName();

    String getNick();

    void setNick(String nick);

    String getPassword();

    void setPassword(String password);

    String getDiscordChannelLinkID();

    void setDiscordChannelLinkID(String discordChannelLinkID);

    Set<String> getWorlds();

    void setWorlds(Set<String> worlds);

    boolean hasWorld(String world);

    boolean hasWorld(World world);

    boolean isBanned(String s);

    boolean isCrossWorld();

    void setCrossWorld(boolean bool);

    boolean isHidden();

    boolean isLocal();

    boolean isMember(Chatter chatter);

    boolean isModerator(String s);

    boolean isMuted(String s);

    boolean isShortcutAllowed();

    void setShortcutAllowed(boolean bool);

    boolean isTransient();

    boolean isVerbose();

    void setVerbose(boolean bool);

    boolean kickMember(Chatter chatter, boolean bool);

    void onFocusGain(Chatter chatter);

    void onFocusLoss(Chatter chatter);

    void processChat(ChannelChatEvent event);

    boolean removeMember(Chatter chatter, boolean announce, boolean flag);

    void removeWorld(String world);

    void setBanned(String chatter, boolean bool);

    void setModerator(String chatter, boolean bool);

    void setMuted(String chatter, boolean bool);

    boolean isMuted();

    void setMuted(boolean bool);

    void sendRawMessage(String message);

    MessageFormatSupplier getFormatSupplier();
}

File: src/main/kotlin/com/fireblazerrr/survivorbot/channel/Channel.kt
package com.fireblazerrr.survivorbot.channel

import com.fireblazerrr.survivorbot.chatter.Chatter
import com.fireblazerrr.survivorbot.spigot.events.ChannelChatEvent
import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.World
import org.bukkit.entity.Player

interface Channel {

    val name: String
    var nick: String
    var format: String
    var password: String
    var discordChannelLinkID: String
    var color: ChatColor
    var distance: Int
    var shortcutAllowed: Boolean
    var verbose: Boolean
    var crossWorld: Boolean
    var muted: Boolean
    val members: MutableSet<Chatter>

    fun addMember(chatter: Chatter, announce: Boolean, flagUpdate: Boolean): Boolean
    fun addWorld(world: String)
    fun announce(message: String)
    fun announce(message: TextComponent)
    fun applyFormat(format: String, message: String, player: String): TextComponent
    fun applyFormat(format: String, message: String, player: Player): TextComponent
    fun applyFormat(format: String, originalFormat: String): String
    fun banMember(chatter: Chatter, announce: Boolean, flag: Boolean): Boolean
    fun emote(chatter: Chatter, message: String)
    fun setBans(chatter: Set<String>)
    fun setModerators(chatters: Set<String>)
    fun setMutes(chatters: Set<String>)
    fun hasWorld(world: World): Boolean
    fun hasWorld(world: String): Boolean
    fun isHidden(): Boolean
    fun isLocal(): Boolean
    fun isChatterMember(name: String): Boolean
    fun isChatterBanned(name: String): Boolean
    fun isChatterModerator(name: String): Boolean
    fun isChatterMuted(name: String): Boolean
    fun isTransient(): Boolean
    fun kickMember(chatter: Chatter, announce: Boolean): Boolean
    fun onFocusGain(chatter: Chatter)
    fun onFocusLoss(chatter: Chatter)
    fun processChat(event: ChannelChatEvent)
    fun removeMember(chatter: Chatter, announce: Boolean, flag: Boolean)
    fun removeWorld(world: String)
    fun setChatterBanned(chatter: String, banned: Boolean)
    fun setChatterModerator(chatter: String, moderator: Boolean)
    fun setChatterMuted(chatter: String, muted: Boolean)
    fun sendRawMessage(message: String)
}